src/persist-client/src/lib.rs

```diff
@@ -446,6 +447,178 @@ mod tests {
         );
     }
 
+    #[tokio::test]
+    async fn invalid_usage() {
+        mz_ore::test::init_logging();
+
+        let data = vec![
+            (("1".to_owned(), "one".to_owned()), 1, 1),
+            (("2".to_owned(), "two".to_owned()), 2, 1),
+            (("3".to_owned(), "three".to_owned()), 3, 1),
+        ];
+
+        let shard_id = ShardId::new();
+        let client = new_test_client().await;
+
+        let (mut write0, read0) = client
+            .expect_open::<String, String, u64, i64>(shard_id)
+            .await;
+
+        write0.expect_compare_and_append(&data, 0, 4).await;
+
+        // InvalidUsage from PersistClient methods.
+        {
+            assert_eq!(
+                format!(
+                    "{:?}",
+                    client.open::<Vec<u8>, String, u64, i64>(shard_id).await
+                ),
+                "Err(CodecMismatch { requested: (\"Vec<u8>\", \"String\", \"u64\", \"i64\"), actual: (\"String\", \"String\", \"u64\", \"i64\") })"
+            );
+            assert_eq!(
+                format!(
+                    "{:?}",
+                    client.open::<String, Vec<u8>, u64, i64>(shard_id).await
+                ),
+                "Err(CodecMismatch { requested: (\"String\", \"Vec<u8>\", \"u64\", \"i64\"), actual: (\"String\", \"String\", \"u64\", \"i64\") })"
+            );
+            assert_eq!(
+                format!(
+                    "{:?}",
+                    client.open::<String, String, i64, i64>(shard_id).await
+                ),
+                "Err(CodecMismatch { requested: (\"String\", \"String\", \"i64\", \"i64\"), actual: (\"String\", \"String\", \"u64\", \"i64\") })"
+            );
+            // We can't test the D param mismatch currently because i64 is literally
+            // the only type that implements both Codec64 and Semigroup right now.
```

Comment:
hmmm this is really surprising to me because Semigroup/Monoid should be implemented for the unsigned integers and i guess it just never was a pressing need

Reply:
opened https://github.com/TimelyDataflow/differential-dataflow/pull/368 which once it merges and we bump differential should let us test the diff param mismatch

Reply:
oh because u64 will implement Semigroup now? nice!